axum/src/json.rs

```diff
@@ -112,6 +113,26 @@ where
     }
 }
 
+impl<T, S> OptionalFromRequest<S> for Json<T>
+where
+    T: DeserializeOwned,
+    S: Send + Sync,
+{
+    type Rejection = JsonRejection;
+
+    async fn from_request(req: Request, state: &S) -> Result<Option<Self>, Self::Rejection> {
+        let is_json = json_content_type(req.headers());
+        let bytes = Bytes::from_request(req, state).await?;
+
+        match (is_json, bytes.is_empty()) {
+            (true, true) => Err(EmptyJsonBody.into()),
+            (true, false) => Ok(Some(Json::from_bytes(&bytes)?)),
+            (false, true) => Ok(None),
+            (false, false) => Err(MissingJsonContentType.into()),
+        }
```

Comment:
No need to check whether the body is empty if there's a JSON content type / introduce a separate rejection variant for that. Also `!is_json` is not the condition I had in mind for returning `None`. Specifically, I only wanted `None` to be returned if there is no `content-type` header at all - not if there is a `content-type` header but it has a value that's not recognized by `json_content_type`.

Reply:
right, let me make some changes and get back to you on this

Reply:
@jplatte made changes, is this what you meant?